tests/e2e/test_snapshot.go

```diff
@@ -0,0 +1,194 @@
+// Copyright © 2023 Cisco Systems, Inc. and/or its affiliates
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package e2e
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/gruntwork-io/terratest/modules/k8s"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+	"github.com/onsi/gomega/format"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+)
+
+type clusterSnapshot struct {
+	resources []metav1.PartialObjectMetadata
+}
+
+func (s *clusterSnapshot) Resources() []metav1.PartialObjectMetadata {
+	return s.resources
+}
+
+// ResourcesAsComparisonType returns a slice of a helper type that makes comparisons easier
+func (s *clusterSnapshot) ResourcesAsComparisonType() []localComparisonPartialObjectMetadataType {
+	var localList []localComparisonPartialObjectMetadataType
+	for _, r := range s.resources {
+		localList = append(localList, localComparisonPartialObjectMetadataType{
+			GVK:       r.GroupVersionKind(),
+			Namespace: r.GetNamespace(),
+			Name:      r.GetName(),
+		})
+	}
+	return localList
+}
+
+// localComparisonPartialObjectMetadataType holds a version of the minimal information required
+// to compare k8s.io/apimachinery/pkg/apis/meta/v1.PartialObjectMetadata instances
+type localComparisonPartialObjectMetadataType struct {
+	GVK       schema.GroupVersionKind
+	Namespace string
+	Name      string
+}
+
+// snapshotCluster takes a clusterSnapshot of a K8s cluster and
+// stores it into the snapshotCluster instance referenced as input
+func snapshotCluster(snapshottedInfo *clusterSnapshot) bool {
+	return When("Get cluster resources state", Ordered, func() {
+		var kubectlOptions k8s.KubectlOptions
+		var err error
+
+		BeforeAll(func() {
+			By("Acquiring K8s config and context")
+			kubectlOptions, err = kubectlOptionsForCurrentContext()
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		var clusterResourceNames []string
+		var namespacedResourceNames []string
+
+		When("Get api-resources names", func() {
+			It("Get cluster-scoped api-resources names", func() {
+				clusterResourceNames, err = listK8sResourceKinds(kubectlOptions, "", "--namespaced=false")
+				Expect(err).NotTo(HaveOccurred())
+				clusterResourceNames = pruneUnnecessaryClusterResourceNames(clusterResourceNames)
```

Comment:
Is it possible and problematic that this `clusterResourceNames` is returned `nil` (it looks like we are using this info to query the resources) ? If so, we might want to do a `Expect(clusterResourceNames).NotTo(BeNil())` after this line

Reply:
As far as pure coding logic is concerned, we could check for nil.

This line `clusterResourceNames = pruneUnnecessaryClusterResourceNames(clusterResourceNames)` makes it so that in fact, when we exit this `It` section, it is never actually nil (because at worst, that `prune...` function returns an empty list that's created during that function).
Also, by the very nature of the request `kubectl api-resources --verbs=list -o name --namespaced=false" doesn't return "nothing" from a working cluster (which we already check for at the beginning of the suite by listing pods and then when acquiring kubeconfig).

If you still feel like it is necessary for the atomic code to be correct, I am ok to add the `nil` check.

Reply:
I would be glad but this can be done in a separate PR.

I understand it's not a necessity due to integration aspects but from a unit standpoint would be better to have it here.

Reply:
Done in e791dd6cdbeef047f00f5801f280b2fcb8cafc1b.

Reply:
> Done in [e791dd6](https://github.com/banzaicloud/koperator/commit/e791dd6cdbeef047f00f5801f280b2fcb8cafc1b).

My typing is too slow :( - didn't get to see the latest commit before I sent out my last reply